californium-core/src/main/java/org/eclipse/californium/core/network/serialization/UdpDataSerializer.java

```diff
@@ -0,0 +1,142 @@
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Matthias Kovatsch - creator and main architect
+ *    Martin Lanter - architect and re-implementation
+ *    Dominique Im Obersteg - parsers and initial implementation
+ *    Daniel Pauli - parsers and initial implementation
+ *    Kai Hudalla - logging
+ *    Bosch Software Innovations GmbH - turn into utility class with static methods only
+ *    Joe Magerramov (Amazon AWS) - CoAP over TCP support
+ ******************************************************************************/
+package org.eclipse.californium.core.network.serialization;
+
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.CODE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.MESSAGE_ID_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_DELTA_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.PAYLOAD_MARKER;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TOKEN_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TYPE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION_BITS;
+
+import java.util.List;
+
+import org.eclipse.californium.core.coap.CoAP.Code;
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.core.coap.Option;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+
+/**
+ * The DataSerialized serializes outgoing messages to byte arrays.
+ */
+public final class UdpDataSerializer implements DataSerializer {
+
+	@Override
+	public byte[] serializeRequest(Request request) {
+		DatagramWriter writer = new DatagramWriter();
+		Code code = request.getCode();
+		serializeMessage(writer, request, code == null ? 0 : code.value);
+		return writer.toByteArray();
+	}
+
+	@Override
+	public byte[] serializeResponse(Response response) {
+		DatagramWriter writer = new DatagramWriter();
+		serializeMessage(writer, response, response.getCode().value);
+		return writer.toByteArray();
+	}
+
+	@Override
+	public byte[] serializeEmptyMessage(Message message) {
+		DatagramWriter writer = new DatagramWriter();
+		serializeMessage(writer, message, 0);
+		return writer.toByteArray();
+	}
+
+	private void serializeMessage(final DatagramWriter writer, final Message message, final int code) {
+		if (message.getToken() == null)
+			throw new NullPointerException("No Token has been set, not even an empty byte[0]");
```

Comment:
The null token should be a violation of the spec, but I'l make
serializers/parsers more defensive in dealing with them.

On Server/Client split: we could combine server/client connectors into a
single implementation that opens outbound connections if needed. We'll need
to be careful to enforce the "responses can only go out over the same
connection that received the request" part of the spec, but I think that's
doable. There might be some other minor complications, but I don't see
anything major.

About the only downside I see is that simple client applications may not
desire to open a listening TCP socket (security considerations and such).
Like you say it's already the case with UDP/DTLS, but there we don't have a
choice.

One alternative is to keep TcpClientConnector as is (only opens outbound
connection), while TcpServerConnector will do both. CoapClient already
creates the right endpoint/connector based on URI scheme. What do you think?

On Tue, Jun 28, 2016 at 8:40 AM Kai Hudalla <notifications@github.com>
wrote:

> In
> californium-core/src/main/java/org/eclipse/californium/core/network/serialization/UdpDataSerializer.java
> <https://github.com/eclipse/californium/pull/61#discussion_r68782740>:
>
> > +	public byte[] serializeResponse(Response response) {
> > +		DatagramWriter writer = new DatagramWriter();
> > +		serializeMessage(writer, response, response.getCode().value);
> > +		return writer.toByteArray();
> > +	}
> > +
> > +	@Override
> > +	public byte[] serializeEmptyMessage(Message message) {
> > +		DatagramWriter writer = new DatagramWriter();
> > +		serializeMessage(writer, message, 0);
> > +		return writer.toByteArray();
> > +	}
> > +
> > +	private void serializeMessage(final DatagramWriter writer, final Message message, final int code) {
> > +		if (message.getToken() == null)
> > +			throw new NullPointerException("No Token has been set, not even an empty byte[0]");
>
> We currently do not explicitly initialize the token in Messages to
> byte[0], thus this check accounts for many NullPointerException when
> running the tests. We should probably change this but until then you should
> better assume a zero-length token when Message.getToken() returns null ...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/californium/pull/61/files/6554875f60f6f9b9c6b35ed2a14fc1f1a6c1d5bf#r68782740>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APu4MhJ9YFTCp4o2Bae7OggHRFPOTOj4ks5qQUBTgaJpZM4I9Q-g>
> .
>